File: c++/Solutions/TwoSum.cpp
#include"TwoSum.hpp"
namespace TwoSum
{

std::vector<int> Solution::twoSum1(std::vector<int>& nums, int target)
{
    std::vector<int> increase {nums};
    std::vector<int> decrease {nums};
    std::vector<int> result;
    std::map<int, int> sumIsTarget;
    std::sort(increase.begin(), increase.end(), [](int a, int b) {return a < b;});
    std::sort(decrease.begin(), decrease.end(), [](int a, int b) {return a > b;});
    for(auto up:increase)
    {
        for(auto down:decrease)
        {
            if (target == up + down &&
                sumIsTarget.find(up) == sumIsTarget.end() &&
                sumIsTarget.find(down) == sumIsTarget.end())
            {
                if ((up == down && 1 < count(nums.begin(),nums.end(),up)) ||
                         up != down)
                {
                        sumIsTarget[up] = down;
                }  
                break;
            }

            if (target > up + down)
            {
                break;
            }
        }
    }

    for(auto element:sumIsTarget)
    {
        int first = element.first;
        int second = element.second;
        int index2 = 0;
        auto fistPosition = std::find(nums.begin(), nums.end(), first);
        auto index1 = std::distance(std::begin(nums), fistPosition);
        if (first != second)
        {
            index2 = std::distance(std::begin(nums), std::find(nums.begin(), nums.end(), second));
        }
        else
        {
            index2 = std::distance(std::begin(nums), std::find(fistPosition+1, nums.end(), second));
        }
        result.push_back(index1);
        result.push_back(index2);
        }
    return result;
    
}

std::vector<int> Solution::twoSum2(std::vector<int>& nums, int target)
{
    std::vector<int> result;
    std::map<int, int> sumIsTarget;
    for(auto element:nums)
    {
        auto second = target - element;
        if (std::find(nums.begin(), nums.end(), second) != nums.end() &&
            (sumIsTarget.find(element) == sumIsTarget.end() &&
            sumIsTarget.find(second) == sumIsTarget.end()))
            {
                if ((second == element && 1 < count(nums.begin(),nums.end(),element)) ||
                    second != element)
                    {
                        sumIsTarget[element] = second;
                    }
            }  
        }

    for(auto element:sumIsTarget)
    {
        int first = element.first;
        int second = element.second;
        int index2 = 0;
        auto fistPosition = std::find(nums.begin(), nums.end(), first);
        auto index1 = std::distance(std::begin(nums), fistPosition);
        if (first != second)
        {
            index2 = std::distance(std::begin(nums), std::find(nums.begin(), nums.end(), second));
        }
        else
            {
            index2 = std::distance(std::begin(nums), std::find(fistPosition+1, nums.end(), second));
            }

        result.push_back(index1);
        result.push_back(index2);
    }
    return result;
}
}

File: python/TwoSum.py
from typing import List
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        increase_list = nums.copy()
        decrease_list = nums.copy()
        increase_list.sort()
        decrease_list.sort(reverse = True)
        sumIsTarget_dict = {}
        result_list = []
        for up in increase_list:
            for down in decrease_list:
                if up + down == target and \
                False == sumIsTarget_dict.__contains__(up) and \
                False == sumIsTarget_dict.__contains__(down):
                    if (up == down and nums.count(up) >1) or \
                    (up != down):
                        sumIsTarget_dict[up] = down
                    break
                if target > up + down:
                    break
        for first, second in sumIsTarget_dict.items():
            indexFirst = nums.index(first)
            indexSecond = 0
            if first != second:
                indexSecond = nums.index(second)
            else:
                indexSecond = nums.index(second,indexFirst+1)
            result_list.append(indexFirst)
            result_list.append(indexSecond)
        
        return result_list
